### DISCUSSION TITLE
DowngradeLevelSetList vs LevelSetList

### DISCUSSION THREAD

**Discussion by ziming** [rectorphp/rector#7344] [Q&A]
Let's say i have a project, some people use newer versions of PHP, some use older.

Which Rector Rule should I use?

DowngradeLevelSetList or LevelSetList?

**Comment by TomasVotruba** [comment]
Hi, 

depends what is your `composer.json` minimal requirement is. Developers of your project should use the same minimal version in `composer.json`. The `LevelSetList` is the best choice in 99 % cases.

The use case for downgrade is mostly for package developers, like ECS, Rector, PHPStan or PHPUnit. This post explains it in a practical easy to understand way https://getrector.org/blog/how-all-frameworks-can-bump-to-php-81-and-you-can-use-older-php

**Reply by ziming** [reply]
I actually read that article before coming here :D

Okay so let's say for 1 of my open source projects, I want to support PHP 7.4 as minimum

but in my own programming I code in PHP 8.1

For me I should use the Downgrade Rector

But what about other random contributors who use say PHP 7.3 ?

Is it okay if i define both Downgrade and LevelSetList in my rector rules array? and if i can which should come first?

As for this part, I wonder if you have an actual github action now on this

```php
 * tag locally
+* downgrade to PHP 8.0
 * push tag to the remote repository
 * let remote repository push the tag to Packagist via git hook
 * let user require a new version in `composer.json.
 * find the tag in Packagist and download the zip from GitHub
```

Thank you very much

**Reply by TomasVotruba** [reply]
> I actually read that article before coming here :D

If that was not enough, these will give you more context :)

* https://getrector.org/blog/how-to-bump-minimal-version-without-leaving-anyone-behind
* https://getrector.org/blog/2021/03/22/rector-010-released-with-php71-support 

> For me I should use the Downgrade Rector

Yes, that's the way to go if you want to develop on PHP 8.1 and release on PHP 7.4 :+1: 

> But what about other random contributors who use say PHP 7.3 ?

You need to agree together which single version to use. Same we apply here in Rector. We develop on PHP 8.1, so any contributor has to run this PHP version to be able to contribute. Nowadays switching PHP version is matter of 2 secs, so after installing desired PHP version, you can start contributing.

**Reply by TomasVotruba** [reply]
> As for this part, I wonder if you have an actual github action now on this

We have exact release flow in the Rector github action here:
https://github.com/rectorphp/rector-src/blob/502d0ad3805984c4e3c0a8c6636c1174068169fc/.github/workflows/build_scoped_rector.yaml#L62-L63
